File: sdk/python/examples/storage_upload_download.py
"""Upload a file then download its bytes back via the 0.3.6 streaming helper.

Demonstrates the StorageService file lifecycle on ``project.storage``:

* ``upload_file`` — RegisterUpload -> HTTP PUT the bytes -> FinalizeUpload.
* ``download_file(file_id, stream=True)`` / ``download_stream`` — stream the
  object bytes back THROUGH the broker via the server-streaming
  ``StorageService.DownloadFile`` RPC (new in 0.3.6), reassembled into the
  full file body. This is the fallback for clients that cannot egress to the
  object store; when egress is available prefer the presigned default
  (``download_file(file_id)`` returns the ``GetDownloadUrlResponse``).
"""

from udb_client import UdbConfig, UdbProject


def main() -> None:
    config = UdbConfig(
        target="127.0.0.1:50051",
        tenant_id="acme",
        project_id="default",
    )

    with UdbProject.connect(config) as project:
        # Login and adopt the canonical tenant/project from the verified
        # principal; installs the bearer across every sub-client.
        project.login_and_adopt_tenant("svc@acme", "password")

        payload = b"hello from the udb python storage example"

        # One-call upload: RegisterUpload -> PUT bytes -> FinalizeUpload.
        finalized = project.storage.upload_file(
            "greeting.txt",
            payload,
            {"content_type": "text/plain", "file_type": "example"},
        )
        file_id = finalized.file_id
        print(f"uploaded file_id={file_id}")

        # Presigned default (no bytes through the broker): returns the
        # GetDownloadUrlResponse carrying a presigned download_url.
        url_response = project.storage.download_file(file_id)
        print(f"presigned download_url={url_response.download_url}")

        # 0.3.6 streaming fallback: stream the bytes back THROUGH the broker.
        downloaded = project.storage.download_file(file_id, stream=True)
        # Equivalent direct call: project.storage.download_stream(file_id)
        assert downloaded == payload
        print(f"downloaded {len(downloaded)} bytes via streaming DownloadFile")


if __name__ == "__main__":
    main()

File: sdk/typescript/examples/storage_upload_download.ts
// Storage upload + streaming download (UDB 0.3.6).
//
// Connect to a UDB broker, log in, upload a file through the native
// StorageService (RegisterUpload -> presigned HTTP PUT -> FinalizeUpload), then
// read the same bytes back via the new 0.3.6 server-streaming
// `StorageService.DownloadFile` RPC.
//
// Run (after building or via ts-node):
//   UDB_TARGET=localhost:50051 \
//   UDB_TENANT=acme \
//   UDB_USER=admin UDB_PASS=secret \
//   node dist/examples/storage_upload_download.js

import { UdbProject } from "@udb_plus/sdk/project";

async function main(): Promise<void> {
  const udb = await UdbProject.connect({
    target: process.env.UDB_TARGET ?? "localhost:50051",
    tenantId: process.env.UDB_TENANT ?? "acme",
    purpose: "storage.example",
    scopes: ["udb:read", "udb:write"],
  });

  try {
    // Password login. `loginAndAdoptTenant` also authenticates the freshly
    // minted bearer and adopts the broker-verified canonical tenant id.
    await udb.loginAndAdoptTenant({
      username: process.env.UDB_USER ?? "admin",
      password: process.env.UDB_PASS ?? "secret",
    });

    // Upload bytes. uploadFile does RegisterUpload -> HTTP PUT to the presigned
    // upload_url -> FinalizeUpload and returns the FinalizeUpload response.
    const payload = Buffer.from("hello from the UDB TypeScript SDK\n", "utf8");
    const finalized: any = await udb.storage.uploadFile("greeting.txt", payload, {
      contentType: "text/plain",
      fileType: "document",
    });
    const fileId: string =
      finalized?.file_id ?? finalized?.file?.file_id ?? "";
    console.log("uploaded file_id:", fileId);

    // Streaming download (0.3.6): pull the raw bytes back over the
    // server-streaming DownloadFile RPC and reassemble into a Uint8Array. Use
    // this when presigned HTTP is unavailable; otherwise prefer the presigned
    // default (`udb.storage.downloadFile(fileId)` -> GetDownloadUrl).
    const bytes: Uint8Array = await udb.storage.downloadFileBytes(fileId);
    console.log("downloaded bytes:", bytes.length);
    console.log("content:", Buffer.from(bytes).toString("utf8"));

    // Equivalent through the canonical accessor with the streaming opt-in:
    //   const bytes = await udb.storage.downloadFile(fileId, { stream: true });
    // The default (no `stream`) instead mints a presigned download URL:
    //   const { download_url } = await udb.storage.downloadFile(fileId);
  } finally {
    udb.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
